internal/observerware/middleware.go

```diff
@@ -22,71 +22,189 @@ package observerware
-	err := h.Handle(ctx, req, w)
-	m.log(ctx, "Handled inbound request.", "unary", req, _timeNow().Sub(start), err)
+	call := m.begin(ctx, transport.Unary, true /* isInbound */, req)
+	wrappedWriter := newWriter(w)
+	err := h.Handle(ctx, req, wrappedWriter)
+	call.End(err, wrappedWriter.isApplicationError)
+	wrappedWriter.free()
 	return err
 }
 
 // Call implements middleware.UnaryOutbound.
 func (m *Middleware) Call(ctx context.Context, req *transport.Request, out transport.UnaryOutbound) (*transport.Response, error) {
-	start := _timeNow()
+	call := m.begin(ctx, transport.Unary, false /* isInbound */, req)
 	res, err := out.Call(ctx, req)
-	m.log(ctx, "Made outbound call.", "unary", req, _timeNow().Sub(start), err)
+
+	isApplicationError := false
+	if res != nil {
+		isApplicationError = res.ApplicationError
+	}
+	call.End(err, isApplicationError)
 	return res, err
 }
 
 // HandleOneway implements middleware.OnewayInbound.
 func (m *Middleware) HandleOneway(ctx context.Context, req *transport.Request, h transport.OnewayHandler) error {
-	start := _timeNow()
+	call := m.begin(ctx, transport.Oneway, true /* isInbound */, req)
 	err := h.HandleOneway(ctx, req)
-	m.log(ctx, "Handled inbound request.", "oneway", req, _timeNow().Sub(start), err)
+	call.End(err, false /* isApplicationError */)
 	return err
 }
 
 // CallOneway implements middleware.OnewayOutbound.
 func (m *Middleware) CallOneway(ctx context.Context, req *transport.Request, out transport.OnewayOutbound) (transport.Ack, error) {
-	start := _timeNow()
+	call := m.begin(ctx, transport.Oneway, false /* isInbound */, req)
 	ack, err := out.CallOneway(ctx, req)
-	m.log(ctx, "Made outbound call.", "oneway", req, _timeNow().Sub(start), err)
+	call.End(err, false /* isApplicationError */)
 	return ack, err
 }
 
-func (m *Middleware) log(ctx context.Context, msg, rpcType string, req *transport.Request, elapsed time.Duration, err error) {
-	if ce := m.logger.Check(zap.DebugLevel, msg); ce != nil {
-		ce.Write(
-			zap.String("rpcType", rpcType),
-			zap.Object("request", req),
-			zap.Duration("latency", elapsed),
-			zap.Bool("successful", err == nil),
-			zap.Error(err),
-			m.extract(ctx),
-		)
+func (m *Middleware) begin(ctx context.Context, rpcType transport.Type, isInbound bool, req *transport.Request) call {
+	now := _timeNow()
+
+	d := newDigester()
+	d.add(req.Caller)
+	d.add(req.Service)
+	d.add(string(req.Encoding))
+	d.add(req.Procedure)
+	d.add(req.ShardKey)
+	d.add(req.RoutingKey)
+	d.add(req.RoutingDelegate)
+	e := m.getOrCreateEdge(d.digest(), req)
+	d.free()
+
+	return call{
+		edge:    e,
+		extract: m.extract,
+		started: now,
+		ctx:     ctx,
+		req:     req,
+		rpcType: rpcType,
+		inbound: isInbound,
+	}
+}
+
+func (m *Middleware) getOrCreateEdge(key []byte, req *transport.Request) *edge {
+	if e := m.getEdge(key); e != nil {
+		return e
 	}
+	return m.createEdge(key, req)
+}
+
+func (m *Middleware) getEdge(key []byte) *edge {
+	m.edgesMu.RLock()
+	e := m.edges[string(key)]
```

Comment:
Based on offline discussion: Can we leave a comment somewhere about why this `[]byte` to `string` conversion doesn't cause an alloc.